driver/src/main/java/org/neo4j/driver/internal/util/InputStreams.java

```diff
@@ -32,18 +33,20 @@
      * @param channel
      * @param into
      */
-    public static void readAll( ReadableByteChannel channel, ByteBuffer into ) throws IOException
+    public static void readAll( ReadableByteChannel channel, ByteBuffer into, Logger logger ) throws IOException
     {
         while( into.remaining() > 0 )
         {
+            int start = into.position();
             int read = channel.read( into );
             if ( read == -1 )
             {
-                throw new ClientException(
+                throw new ClientException( String.format(
                         "Connection terminated while receiving data. This can happen due to network " +
-                        "instabilities, or due to restarts of the database. Expected " + into.limit() + "bytes, " +
-                        "recieved " + BytePrinter.hex( into ) + "." );
+                        "instabilities, or due to restarts of the database. Expected %s bytes, received %s.",
+                        into.limit(), BytePrinter.hex( into ) ) );
             }
+            logger.trace( "S: " + BytePrinter.hex( into, start, read ) );
```

Comment:
Same as with the channel thing - if we put the logging on the channel level, we wont need this